### PR TITLE
MRG: maintain_fix and more custom calibration

### PR DESCRIPTION
**Pull request by maddycapp27**
This adds a function to track fixation over time called ``maintain_fix`` which makes sure you're looking in the correct area at least once per interval.

Additionally, I've added a option to have a truly custom calibration with specified coordinates.

Both still need error checking and tests.

### CODE REVIEW

File: expyfun/_eyelink_controller.py [hunk 1]
```diff
@@ -615,9 +615,17 @@ def custom_calibration(self, ctype='HV5', horiz=2./3., vert=2./3.,
...
+            if not coordinates == None:
+                raise ValueError('If ctype is not \'custom\' coordinates canno'
+                                 't be used to generate calibration pattern.')
+        else:
+            if isinstance(coordinates, list):
+                raise ValueError('Coordinates must be in a list.')
+            if any([not isinstance(c, list) or len(c) != 2 
+                    for c in coordinates]):
+                raise ValueError('Each coordinate must be a list with length 2'
+                                '.')
```
Comment (maddycapp27) [inline expyfun/_eyelink_controller.py]: Should there also be a stipulation that there should be a minimum number of coordinates? I'm assuming that the eyetracker can't do much with less than 3.

**Comment by maddycapp27** [comment]
@larsoner @drammock @rkmaddox 

This also seems to be an unrelated AppVeyor failure. Not sure what to do about coveralls since I'm getting this error "The file "expyfun/_eyelink_controller.py" isn't available on github. Either it's been removed, or the repo root directory needs to be updated." when I try to open it. Regardless I think I'm ready for review on this one as well.

File: expyfun/_eyelink_controller.py [hunk 2]
```diff
@@ -528,30 +528,105 @@ def wait_for_fix(self, fix_pos, fix_time=0., tol=100., max_wait=np.inf,
...
+        `period` seconds.
+
+        Parameters
+        ----------
+        fix_pos : tuple (length 2)
+            The screen position (in specified units) required.
+        check_duration : float
+            Total amount of time to check.
+        tol : float
+            The tolerance (in pixels) to consider the target hit.
```
Comment (larsoner) [inline expyfun/_eyelink_controller.py]: We usually allow units for these sorts of things. Maybe this should be in the same units as `fix_pos`. And is it a radius (circular area) or w/h (square)? (I should be able to tell from the var name ideally, and if not there, then the description.)

File: expyfun/_eyelink_controller.py [hunk 3]
```diff
@@ -528,30 +528,105 @@ def wait_for_fix(self, fix_pos, fix_time=0., tol=100., max_wait=np.inf,
...
         if ctype not in allowed_types:
             raise ValueError('ctype cannot be "{0}", but must be one of {1}'
                              ''.format(ctype, allowed_types))
+        if ctype != 'custom':
+            if coordinates is not None:
+                raise ValueError('If ctype is not \'custom\' coordinates canno'
+                                 't be used to generate calibration pattern.')
+        else:
+            if isinstance(coordinates, list):
+                raise ValueError('Coordinates must be in a list.')
```
Comment (larsoner) [inline expyfun/_eyelink_controller.py]: if it's a list, you raise an error that it must be a list ... ?

File: expyfun/_eyelink_controller.py [hunk 4]
```diff
@@ -528,30 +528,105 @@ def wait_for_fix(self, fix_pos, fix_time=0., tol=100., max_wait=np.inf,
...
+            if any([not isinstance(c, list) or len(c) != 2
+                    for c in coordinates]):
+                raise ValueError('Each coordinate must be a list with length 2'
+                                 '.')
```
Comment (larsoner) [inline expyfun/_eyelink_controller.py]: The easier check is to do `coordinates = np.array(coordinates, float)` and then check `coordinates.shape`

File: expyfun/_eyelink_controller.py (hunk 3 above)
Comment (maddycapp27) [reply expyfun/_eyelink_controller.py]: OOPS!

File: expyfun/_eyelink_controller.py (hunk 4 above)
Comment (maddycapp27) [reply expyfun/_eyelink_controller.py]: ah ok, makes sense

File: expyfun/_eyelink_controller.py (hunk 2 above)
Comment (maddycapp27) [reply expyfun/_eyelink_controller.py]: I was using ``wait_for_fix()`` as a template which has allows any units for fix_pos but _only_ does tolerance in pixels.

Comment (maddycapp27) [reply expyfun/_eyelink_controller.py]: Do we want to change both ``wait_for_fix()`` and ``maintain_fix()`` to allow for any units?

Comment (larsoner) [reply expyfun/_eyelink_controller.py]: Hmm, that's a good question. Does it seem worth it to you?

Comment (maddycapp27) [reply expyfun/_eyelink_controller.py]: Well, the issue is that the tolerances in pixels make sense anywhere on the screen but if you have fixation dots off to the side and are trying to use degrees as tolerances, things get a little funky. So if we allowed for any units I'd say we'd end up converting whatever tolerances in degrees or norm to pixels at the center and then use those.

I think it's worth it as long as we're transparent about the units not being true on the sides.

Comment (larsoner) [reply expyfun/_eyelink_controller.py]: Good point. We can keep pixels for now, then, and add units later if we need them

**Comment by larsoner** [comment]
Something is wrong with your development workflow, you are pushing commits as @rkmaddox

**Comment by maddycapp27** [comment]
Pushing as @rkmaddox because I'm on the behavioral computer in the lab rather than at my desktop.

**Comment by larsoner** [comment]
It would be good to figure out a better system in the not too distant future, otherwise things like `git blame` don't work correctly.

File: expyfun/_experiment_controller.py [hunk 5]
```diff
@@ -1069,6 +1069,10 @@ def _log_presses(self, pressed):
     def check_force_quit(self):
         """Check to see if any force quit keys were pressed."""
         self._response_handler.check_force_quit()
+    
+    def check_interrupts(self):
+        """Check to see if any interrupt keys were pressed."""
+        self._response_handler.check_interrupts()
```
Comment (larsoner) [inline expyfun/_experiment_controller.py]: Shouldn't be part of this PR, right?

Comment (maddycapp27) [reply expyfun/_experiment_controller.py]: Yes, that is my mistake

Comment (maddycapp27) [reply expyfun/_experiment_controller.py]: (serves me right for trying to multitask)

File: expyfun/_eyelink_controller.py [hunk 6]
```diff
@@ -528,30 +528,102 @@ def wait_for_fix(self, fix_pos, fix_time=0., tol=100., max_wait=np.inf,
...
         if ctype not in allowed_types:
             raise ValueError('ctype cannot be "{0}", but must be one of {1}'
                              ''.format(ctype, allowed_types))
+        if ctype != 'custom':
+            if coordinates is not None:
+                raise ValueError('If ctype is not \'custom\' coordinates canno'
+                                 't be used to generate calibration pattern.')
+        else:
+            if not isinstance(coordinates, list):
+                raise ValueError('Coordinates must be in a list.')
```
Comment (larsoner) [inline expyfun/_eyelink_controller.py]: you don't need this check, `np.array` is good enough (you should really allow list, tuple, ndarray, etc.)

Comment (larsoner) [reply expyfun/_eyelink_controller.py]: (also this is redundant with the check that is done later)

File: expyfun/_eyelink_controller.py [hunk 7]
```diff
@@ -573,6 +645,13 @@ def custom_calibration(self, ctype='HV5', horiz=2./3., vert=2./3.,
             mat = np.array([[0, 0], [1, 0], [-1, 0], [0, 1], [0, -1], [1, 1],
                             [-1, -1], [1, -1], [-1, 1], [.5, .5], [-.5, -.5],
                             [.5, -.5], [-.5, .5]])
+        elif ctype == 'custom':
+            if not isinstance(coordinates, list):
+                raise ValueError('Coordinates must be in a list.')
```
Comment (larsoner) [inline expyfun/_eyelink_controller.py]: you don't need this check

File: expyfun/_eyelink_controller.py [hunk 8]
```diff
@@ -573,6 +645,13 @@ def custom_calibration(self, ctype='HV5', horiz=2./3., vert=2./3.,
...
+            mat = np.array(coordinates, float)
+            if mat.shape[-1] != 2:
```
Comment (larsoner) [inline expyfun/_eyelink_controller.py]: `if mat.ndim != 2 or mat.shape[-1] != 2`

**Comment by drammock** [comment]
I don't have time for an in-depth review right now, but running `make test` locally, it fails flake:

```
Running flake8
expyfun/_eyelink_controller.py:647:15: E271 multiple spaces after keyword
expyfun/_input_controllers.py:170:51: W291 trailing whitespace
```

I also get some warnings (these may not be new):
```
Test EC methods. ... /home/drmccloy/Builds/expyfun/expyfun/_experiment_controller.py:1380: RuntimeWarning: invalid value encountered in multiply
  samples *= self._stim_scaler
/home/drmccloy/Builds/expyfun/expyfun/_sound_controllers.py:151: RuntimeWarning: invalid value encountered in less
  data[data < -1] = -1
/home/drmccloy/Builds/expyfun/expyfun/_sound_controllers.py:152: RuntimeWarning: invalid value encountered in greater
  data[data > 1] = 1
```

**Comment by larsoner** [comment]
Yeah those warnings are old

**Comment by larsoner** [comment]
I'll go ahead and merge because it's just OSX we're waiting for, thanks @maddycapp27
